### PR TITLE
feat: self-hosted real-time collab and presence

### PR DESCRIPTION
**Pull request by tangjeff0**

### CODE REVIEW

File: shadow-cljs.edn [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {:lein   true
- :nrepl {:port 8777}
+ :nrepl {:port 8778}
```
Comment (arkRedM) [inline shadow-cljs.edn:2]: revert

File: shadow-cljs.edn [hunk 2]
```diff
@@ -17,7 +17,7 @@
                 :devtools        {:preloads [devtools.preload
                                              day8.re-frame-10x.preload]
                                   :http-root    "resources/public"
-                                  :http-port    3000}}
+                                  :http-port    3001}}
```
Comment (arkRedM) [inline shadow-cljs.edn:20]: revert

**Comment by arkRedM** [comment]
Next steps
1. Basic -- lint, style
2. a. Clarification: For auth, a token and ip shared with folks is enough (username and pass doesn't add much) 
    b. Add option to switch between connection to remote graph and local graph.
3. One-click deploy to  ___? (heroku | some other prov?)

**Review by arkRedM — CHANGES_REQUESTED** [review]

**Review by agentydragon — CHANGES_REQUESTED** [review]
Nice work. I think it might take a bit more to get it ready.

Couple high level things:

- I believe we ended up deciding to put the server into the same repo, right? Can you add the server into the PR?

- Since it might take a bit to get into good state, I'd recommend submitting separately whatever can be submitted separately. For example, the display name stuff and maybe the presence Reagent component could be submitted (maybe with tests, too) separately. Code can also be behind a feature flag (though I believe Athens doesn't right now have feature flags...)

File: src/cljs/athens/views/settings_page.cljs [hunk 3]
```diff
@@ -150,5 +151,15 @@
...
+             [:span (str "Athens will save and create a local backup " @debounce-save-time! " seconds after your last edit.")])]]
+
+         [:div {:style {:margin "20px 0"}}
+          [:h5 "Your Name"]
+          [:div {:style {:display "flex" :justify-content "space-between"
+                         :margin "10px 0"}}
+           [:input {:style {:width "12em"}
+                    :value (:name @(subscribe [:user/current]))
+                    :on-change #(do (dispatch [:user/set :name (.. % -target -value)])
+                                    (js/localStorage.setItem "user/name" (.. % -target -value)))}]
```
[the remark this answers is not in the document]
Comment (agentydragon) [reply src/cljs/athens/views/settings_page.cljs:163]: https://github.com/Day8/re-frame/blob/master/docs/Effects.md says re-frame has built-in effects for setting local store stuff

https://github.com/Day8/re-frame/blob/master/docs/Effects.md#the-effects-map

File: project.clj [hunk 4]
```diff
@@ -30,7 +30,9 @@
                  [garden "1.3.10"]
                  [stylefy "2.2.0"]
                  [tick "0.4.26-alpha"]
-                 [com.rpl/specter "1.1.3"]]
+                 [com.rpl/specter "1.1.3"]
+                 [com.taoensso/sente "1.16.2"]
+                 [datsync "0.0.1-alpha2-SNAPSHOT"]]
```
[the remark this answers is not in the document]
Comment (tangjeff0) [reply project.clj:35]: Used in `src/cljs/athens/datsync_utils.cljs`

**Review by tangjeff0 — COMMENTED** [review]
Some thoughts:

- Reconnection doesn't seem to work if a computer sleeps and awakens
- Definitely need to move db picker into settings, getting crowded now. CC @shanberg 
- `You` in top right always seems to say `You`. Seems like there should be other user names if they are on the same page.
image
- Seems like the ability to "Follow" a user while they navigate to other pages or see what page they are on at the moment would be nice.
- Need to click on the People icon to see if the name of the User that is there. Hover probably makes more sense
image

**Comment by agentydragon** [comment]
> Seems like the ability to "Follow" a user while they navigate to other pages or see what page they are on at the moment would be nice.

Agree, but I think this would be better left for later, to make the initial version smaller.

File: project.clj (hunk 4 above)
[the remark this answers is not in the document]
Comment (agentydragon) [reply project.clj:35]: ah, you're right.

but then, still, depending on a SNAPSHOT version does not sound to me like a good idea. could we

File: src/cljs/athens/views/presence.cljs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+                                      (filter #(and ctx-uid
+                                                    (or (= (:current/uid %) ctx-uid)
+                                                        (= (:editing/uid %) ctx-uid))
+                                                    (not= (:random/id %)
+                                                          ws/cur-random))))
+         others-in-cur-uid-count (count others-in-cur-uid)
+         total-in-cur-uid (->> curr-presence vals
+                               (filter #(and ctx-uid
+                                             (or (= (:current/uid %) ctx-uid)
+                                                 (= (:editing/uid %) ctx-uid))))
```
[the remark this answers is not in the document]
Comment (agentydragon) [reply src/cljs/athens/views/presence.cljs]: also the predicate shares a bunch with lines 43..47

**Comment by arkRedM** [comment]
> Pressing enter at the same time under the same parent
This should be done a little cleaner I agree. Too many bad nodes are not ideal.

> Reconnection doesn't seem to work if a computer sleeps and awakens
This is probably a must. I think I should come up with a bare min reconnection logic.

**Comment by arkRedM** [comment]
> You in top right always seems to say You. Seems like there should be other user names if they are on the same page.

it should come up as "You" or "You and n others". Will look into it

**Comment by tangjeff0** [comment]
Btw, I'm not exactly sure what your timeline looks like right now, so I've been working off of your branch @arkRedM in case you're availability is limited/variable: https://github.com/arkRedM/athens/pull/1

**Comment by tangjeff0** [comment]
Getting this error when deleting block(s):

image

**Comment by tangjeff0** [comment]
@arkRedM might be good to merge/rebase  what I've done in on my `presence` branch. Some small refactors and fixes.

**Comment by arkRedM** [comment]
> @arkRedM might be good to merge/rebase what I've done in on my `presence` branch. Some small refactors and fixes.

Oh! I did the exact same changes 🙆‍♂️(with slight differences)

**Comment by arkRedM** [comment]
Okay too many commits and diffs, easier thing is I go over commits from your branch one by one and add the final version. @tangjeff0  how does that sound?

**Comment by arkRedM** [comment]
> Okay too many commits and diffs, easier thing is I go over commits from your branch one by one and add the final version. @tangjeff0 how does that sound?

Went ahead and did that(took credit for your commits :P)

**Comment by tangjeff0** [comment]
Would be ideal if there was documentation on how this was architected. https://github.com/athensresearch/athens-backend should also have setup instructions, including how to migrate from local datascript `index.transit` to datahike for our enterprise customers.

**Comment by arkRedM** [comment]
> setup instructions
Figuring out some parts will add it once everything is clear 😅

**Pull request merged**